pkg/services/ngalert/state/historian/annotation.go

```diff
@@ -31,48 +31,82 @@ func NewAnnotationHistorian(annotations annotations.Repository, dashboards dashb
-		NewState:  currentData.String(),
-		Text:      annotationText,
-		Epoch:     evaluatedAt.UnixNano() / int64(time.Millisecond),
-	}
+func (h *AnnotationStateHistorian) RecordStates(ctx context.Context, states []state.ContextualState) {
+	// Build annotations before starting goroutine, to make sure all data is copied and won't mutate underneath us.
+	annotations := h.buildAnnotations(states)
+	go h.recordAnnotationsSync(ctx, annotations)
+}
 
-	dashUid, ok := rule.Annotations[ngmodels.DashboardUIDAnnotation]
-	if ok {
-		panelUid := rule.Annotations[ngmodels.PanelIDAnnotation]
+type itemWithMetadata struct {
+	annotations.Item
+	dashUID       string
+	parsedPanelID int64
+}
 
-		panelId, err := strconv.ParseInt(panelUid, 10, 64)
-		if err != nil {
-			h.log.Error("error parsing panelUID for alert annotation", "panelUID", panelUid, "alertRuleUID", rule.UID, "err", err.Error())
-			return
-		}
+func (h *AnnotationStateHistorian) buildAnnotations(states []state.ContextualState) []itemWithMetadata {
+	items := make([]itemWithMetadata, 0, len(states))
+	for _, state := range states {
+		h.log.Debug("alert state changed creating annotation", "alertRuleUID", state.AlertRuleUID, "newState", state.Formatted(), "oldState", state.PreviousFormatted())
 
-		query := &models.GetDashboardQuery{
-			Uid:   dashUid,
-			OrgId: rule.OrgID,
+		labels := removePrivateLabels(state.Labels)
+		annotationText := fmt.Sprintf("%s {%s} - %s", state.Rule.Title, labels.String(), state.Formatted())
+
+		item := itemWithMetadata{
+			Item: annotations.Item{
+				AlertId:   state.Rule.ID,
+				OrgId:     state.OrgID,
+				PrevState: state.PreviousFormatted(),
+				NewState:  state.Formatted(),
+				Text:      annotationText,
+				Epoch:     state.LastEvaluationTime.UnixNano() / int64(time.Millisecond),
+			},
 		}
 
-		err = h.dashboards.GetDashboard(ctx, query)
-		if err != nil {
-			h.log.Error("error getting dashboard for alert annotation", "dashboardUID", dashUid, "alertRuleUID", rule.UID, "err", err.Error())
-			return
+		dashUID, ok := state.Annotations[ngmodels.DashboardUIDAnnotation]
+		if ok {
+			panelUID := state.Annotations[ngmodels.PanelIDAnnotation]
+
+			panelID, err := strconv.ParseInt(panelUID, 10, 64)
+			if err != nil {
+				h.log.Error("error parsing panelUID for alert annotation", "panelUID", panelUID, "alertRuleUID", state.AlertRuleUID, "err", err.Error())
+				continue
+			}
+			item.dashUID = dashUID
+			item.parsedPanelID = panelID
 		}
+		items = append(items, item)
+	}
+	return items
+}
+
+func (h *AnnotationStateHistorian) recordAnnotationsSync(ctx context.Context, items []itemWithMetadata) {
+	annotations := make([]annotations.Item, 0, len(items))
+
+	for _, item := range items {
+		if item.dashUID != "" {
+			query := &models.GetDashboardQuery{
+				Uid:   item.dashUID,
+				OrgId: item.Item.OrgId,
+			}
 
-		item.PanelId = panelId
-		item.DashboardId = query.Result.Id
+			err := h.dashboards.GetDashboard(ctx, query)
```

Comment:
all states\items in this method belong to the same rule. I think we should optimize it to not query dashboard for each item.

Reply:
_technically_ the interface is more generic than that, and can take many states from multiple rules.

But, in practice it's never used like that. In the meantime we can adjust the interface to record states for a specific rule and take the rule as a separate arg. Then, the optimization you describe happens almost automatically since there is only ever one annotation set.

My next PR also improves this, I'm caching this query in memory